src/theory/arith/nl/icp/icp_solver.cpp

```diff
@@ -0,0 +1,364 @@
+  {
+    if (!d_state.d_assignment.has(vars.second)) continue;
+    Node v = vars.first;
+    poly::Interval i = d_state.d_assignment.get(vars.second);
+    if (!is_minus_infinity(get_lower(i)))
+    {
+      Kind rel = get_lower_open(i) ? Kind::GT : Kind::GEQ;
+      Node c = nm->mkNode(rel, v, value_to_node(get_lower(i), v));
+      if (!d_state.d_origins.isInOrigins(v, c))
+      {
+        Node premise = d_state.d_origins.getOrigins(v);
+        Trace("nl-icp") << premise << " => " << c << std::endl;
+        Node lemma = Rewriter::rewrite(nm->mkNode(Kind::IMPLIES, premise, c));
+        if (lemma.isConst())
+        {
+          Assert(lemma == nm->mkConst<bool>(true));
+        }
+        else
+        {
+          Trace("nl-icp") << "Adding lemma " << lemma << std::endl;
+          lemmas.emplace_back(lemma);
+        }
+      }
+    }
+    if (!is_plus_infinity(get_upper(i)))
+    {
+      Kind rel = get_upper_open(i) ? Kind::LT : Kind::LEQ;
+      Node c = nm->mkNode(rel, v, value_to_node(get_upper(i), v));
+      if (!d_state.d_origins.isInOrigins(v, c))
+      {
+        Node premise = d_state.d_origins.getOrigins(v);
+        Trace("nl-icp") << premise << " => " << c << std::endl;
+        Node lemma = Rewriter::rewrite(nm->mkNode(Kind::IMPLIES, premise, c));
+        if (lemma.isConst())
+        {
+          Assert(lemma == nm->mkConst<bool>(true));
+        }
+        else
+        {
+          Trace("nl-icp") << "Adding lemma " << lemma << std::endl;
+          lemmas.emplace_back(lemma);
+        }
+      }
+    }
+  }
+  return lemmas;
+}
+
+void ICPSolver::reset(const std::vector<Node>& assertions)
+{
+  d_state.reset();
+  for (const auto& n : assertions)
+  {
+    Node tmp = Rewriter::rewrite(n);
+    Trace("nl-icp") << "Adding " << tmp << std::endl;
+    if (tmp.getKind() != Kind::CONST_BOOLEAN)
+    {
+      if (!d_state.d_bounds.add(tmp))
+      {
+        addCandidate(tmp);
+      }
+    }
+  }
+}
+
+void ICPSolver::check()
+{
+  initOrigins();
+  d_state.d_assignment = d_state.d_bounds.get();
+  bool did_progress = false;
+  bool progress = false;
+  do
+  {
+    switch (doPropagationRound())
+    {
+      case icp::PropagationResult::NOT_CHANGED: progress = false; break;
+      case icp::PropagationResult::CONTRACTED:
+      case icp::PropagationResult::CONTRACTED_STRONGLY:
+      case icp::PropagationResult::CONTRACTED_WITHOUT_CURRENT:
+      case icp::PropagationResult::CONTRACTED_STRONGLY_WITHOUT_CURRENT:
+        did_progress = true;
+        progress = true;
+        break;
+      case icp::PropagationResult::CONFLICT:
+        Trace("nl-icp") << "Found a conflict: " << d_state.d_conflict
+                        << std::endl;
+
+        d_im.addConflict(d_state.d_conflict, InferenceId::NL_ICP_CONFLICT);
+        did_progress = true;
+        progress = false;
+        break;
+    }
+  } while (progress);
+  if (did_progress)
+  {
+    for (const auto& l : generateLemmas())
```

Comment:
This loop is really crazy to me :)

I guess this is assuming a fixed call to generateLemmas() and then iterates over what is returned?  Regardless can you cache a reference and iterate?

Reply:
Tbh, I really consider this to be better / more readable than the alternative:
```suggestion
    std::vector<Node> lemmas = generateLemmas();
    for (const auto& l : lemmas)
```
If you really prefer it, I'll change :-)

Reply:
Yes, I really prefer it :)